drivers/scheduler/anthos/anthos.go

```diff
@@ -0,0 +1,576 @@
+func (anth *anthos) UpgradeScheduler(version string) error {
+	log.Info("Upgrading Anthos user cluster")
+	if !versionReg.MatchString(version) {
+		return fmt.Errorf("incorrect upgrade version: [%s] is provided", version)
+	}
+	if err := anth.verifyUpgradeVersion(version); err != nil {
+		return err
+	}
+	if err := anth.loadInstances(); err != nil {
+		return err
+	}
+	if err := downloadAndInstallGsutils(); err != nil {
+		return err
+	}
+	if err := anth.upgradeAdminWorkstation(version); err != nil {
+		return err
+	}
+	startTime := time.Now()
+	if err := anth.upgradeUserCluster(version); err != nil {
+		return err
+	}
+	timeTaken := time.Since(startTime)
+	log.Infof("Anthos user cluster took: %v time to complete the upgrade", timeTaken)
+	if err := anth.RefreshNodeRegistry(); err != nil {
+		return err
+	}
+	if err := anth.invokeUpgradeAdminCluster(version); err != nil {
+		return err
+	}
+	return nil
+}
+
+// invokeUpgradeAdminCluster start admin cluster upgrade
+func (anth *anthos) invokeUpgradeAdminCluster(version string) error {
+	log.Info("Upgrading admin cluster")
+	initTime := time.Now()
+	if err := anth.upgradeAdminCluster(version); err != nil {
+		return err
+	}
+	timeTaken := time.Since(initTime)
+	log.Infof("Anthos upgrade took: %v time to complete upgrade from % to %s version",
+		timeTaken, anth.version, version)
+	if err := anth.updateNodeInstance(); err != nil {
+		return err
+	}
+	if err := anth.saveInstance(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// updateNodeInstance will update the host info after upgrade
+func (anth *anthos) updateNodeInstance() error {
+	log.Info("Updating node Instance")
+	var startAdminNodeIndex int = 2
+	var lastAdminNodeIndex int = 4
+	k8sOps, err := core.NewInstanceFromConfigFile(path.Join(anth.confPath, kubeConfig))
+	if err != nil {
+		return err
+	}
+	adminNodeList, err := k8sOps.GetNodes()
+	if err != nil {
+		return err
+	}
+	adminIndex := startAdminNodeIndex
+	for _, adminNode := range adminNodeList.Items {
+		if adminIndex > lastAdminNodeIndex {
+			break
+		}
+		anth.instances[adminIndex].HostName = adminNode.Name
+		anth.instances[adminIndex].PublicIpAddress = adminNode.Status.Addresses[0].Address
+		anth.instances[adminIndex].PrivateIpAddress = adminNode.Status.Addresses[0].Address
+		adminIndex += 1
+	}
+	return nil
+}
+
+// saveInstance save the instances.json after upgrade
+func (anth *anthos) saveInstance() error {
+	log.Debug("Saving instances.json")
+	b, err := json.MarshalIndent(anth.instances, "", "  ")
+	if err != nil {
+		return fmt.Errorf("cannot marshall Instances, %v", err)
+	}
+	if os.WriteFile(path.Join(anth.instPath, jsonInstances), b, 0644); err != nil {
+		return fmt.Errorf("couldn't write to %s/%s: %v", anth.instPath, jsonInstances, err)
+	}
+	return nil
+}
+
+// verifyUpgradeVersion validates that correct version is provided for upgrade or not
+func (anth *anthos) verifyUpgradeVersion(upgradeVersion string) error {
```

Comment:
can we move this verification as part of test

Reply:
This method is needed to invoke as part of upgradeScheduler method. This need to called before upgrading Anthos !!